File: pkg/analyzers/halstead/formatter.go
package halstead

import (
	"encoding/json"
	"fmt"
	"io"

	"gopkg.in/yaml.v3"

	"github.com/Sumatoshi-tech/codefang/pkg/analyzers/analyze"
	"github.com/Sumatoshi-tech/codefang/pkg/analyzers/common"
)

// Configuration constants for Halstead report formatting.
const (
	// MaxItemsValue is the maximum number of items displayed in Halstead report summaries.
	MaxItemsValue           = 10
	bugsThresholdMedium     = 0.5
	bugsThresholdMin        = 0.1
	difficultyThresholdHigh = 5
	effortThresholdHigh     = 1000
	magic10000              = 10000
	magic15                 = 15
)

// ReportFormatter handles formatting of Halstead analysis reports.
type ReportFormatter struct {
	reporter *common.Reporter
}

// NewReportFormatter creates a new report formatter.
func NewReportFormatter() *ReportFormatter {
	config := common.ReportConfig{
		Format:         "text",
		IncludeDetails: true,
		SortBy:         "volume",
		SortOrder:      "desc",
		MaxItems:       MaxItemsValue,
		MetricKeys:     []string{"volume", "difficulty", "effort", "time_to_program", "delivered_bugs"},
	}

	return &ReportFormatter{
		reporter: common.NewReporter(config),
	}
}

// FormatReport formats the analysis report for display.
func (rf *ReportFormatter) FormatReport(report analyze.Report, w io.Writer) error {
	formatted, err := rf.reporter.GenerateReport(report)
	if err != nil {
		return err
	}

	_, err = fmt.Fprint(w, formatted)
	if err != nil {
		return fmt.Errorf("formatreport: %w", err)
	}

	return nil
}

// FormatReportJSON formats the analysis report as JSON.
func (rf *ReportFormatter) FormatReportJSON(report analyze.Report, w io.Writer) error {
	metrics, err := ComputeAllMetrics(report)
	if err != nil {
		metrics = &ComputedMetrics{}
	}

	jsonData, err := json.MarshalIndent(metrics, "", "  ")
	if err != nil {
		return fmt.Errorf("formatreportjson: %w", err)
	}

	_, err = fmt.Fprint(w, string(jsonData))
	if err != nil {
		return fmt.Errorf("formatreportjson: %w", err)
	}

	return nil
}

// FormatReportYAML formats the analysis report as YAML.
func (rf *ReportFormatter) FormatReportYAML(report analyze.Report, w io.Writer) error {
	metrics, err := ComputeAllMetrics(report)
	if err != nil {
		metrics = &ComputedMetrics{}
	}

	data, err := yaml.Marshal(metrics)
	if err != nil {
		return fmt.Errorf("formatreportyaml: %w", err)
	}

	_, err = w.Write(data)
	if err != nil {
		return fmt.Errorf("formatreportyaml: %w", err)
	}

	return nil
}

// GetHalsteadMessage returns a message based on the Halstead metrics.
func (rf *ReportFormatter) GetHalsteadMessage(volume, difficulty, effort float64) string {
	if volume <= 100 && difficulty <= 5 && effort <= 1000 {
		return "Excellent complexity - code is simple and maintainable"
	}

	if volume <= 1000 && difficulty <= 15 && effort <= 10000 {
		return "Good complexity - code is reasonably complex"
	}

	if volume <= 5000 && difficulty <= 30 && effort <= 50000 {
		return "Fair complexity - consider simplifying some functions"
	}

	return "High complexity - code should be refactored for better maintainability"
}

// GetVolumeAssessment returns an assessment with emoji for volume.
func (rf *ReportFormatter) GetVolumeAssessment(volume float64) string {
	if volume <= volumeThresholdHigh {
		return "🟢 Low"
	}

	if volume <= magic1000 {
		return "🟡 Medium"
	}

	return "🔴 High"
}

// GetDifficultyAssessment returns an assessment with emoji for difficulty.
func (rf *ReportFormatter) GetDifficultyAssessment(difficulty float64) string {
	if difficulty <= difficultyThresholdHigh {
		return "🟢 Simple"
	}

	if difficulty <= magic15 {
		return "🟡 Moderate"
	}

	return "🔴 Complex"
}

// GetEffortAssessment returns an assessment with emoji for effort.
func (rf *ReportFormatter) GetEffortAssessment(effort float64) string {
	if effort <= effortThresholdHigh {
		return "🟢 Low"
	}

	if effort <= magic10000 {
		return "🟡 Medium"
	}

	return "🔴 High"
}

// GetBugAssessment returns an assessment with emoji for delivered bugs.
func (rf *ReportFormatter) GetBugAssessment(bugs float64) string {
	if bugs <= bugsThresholdMin {
		return "🟢 Low Risk"
	}

	if bugs <= bugsThresholdMedium {
		return "🟡 Medium Risk"
	}

	return "🔴 High Risk"
}

File: rust/crates/cf-halstead/src/formatter.rs
//! Report formatting and assessment helpers (`formatter.go`).
//!
//! The JSON/YAML/binary report bodies serialize the [`ComputedMetrics`] view
//! (NOT the raw analyzer report) via the Go-byte-compatible encoders, matching
//! `FormatReportJSON` (json.MarshalIndent two-space), `FormatReportYAML`
//! (yaml.Marshal) and `FormatReportBinary` (CFB1 envelope around compact JSON).

use std::io::{self, Write};

use crate::gojson;
use crate::metrics::ComputedMetrics;

// --- Assessment thresholds (formatter.go) ---

const VOLUME_THRESHOLD_HIGH: f64 = 5000.0; // volumeThresholdHigh
const MAGIC_1000: f64 = 1000.0;
const DIFFICULTY_THRESHOLD_HIGH: f64 = 5.0;
const MAGIC_15: f64 = 15.0;
const EFFORT_THRESHOLD_HIGH: f64 = 1000.0;
const MAGIC_10000: f64 = 10000.0;
const BUGS_THRESHOLD_MIN: f64 = 0.1;
const BUGS_THRESHOLD_MEDIUM: f64 = 0.5;

/// Formats Halstead analysis reports (`ReportFormatter`).
#[derive(Debug, Clone, Copy, Default)]
pub struct ReportFormatter;

impl ReportFormatter {
    /// Creates a new formatter.
    #[must_use]
    pub fn new() -> Self {
        Self
    }

    /// Overall message from the file-level measures (`GetHalsteadMessage`).
    ///
    /// Note: this is distinct from the aggregator's volume-only message
    /// (`buildHalsteadMessage`); it is the one stored in the per-file report.
    #[must_use]
    pub fn halstead_message(&self, volume: f64, difficulty: f64, effort: f64) -> &'static str {
        if volume <= 100.0 && difficulty <= 5.0 && effort <= 1000.0 {
            return "Excellent complexity - code is simple and maintainable";
        }
        if volume <= 1000.0 && difficulty <= 15.0 && effort <= 10000.0 {
            return "Good complexity - code is reasonably complex";
        }
        if volume <= 5000.0 && difficulty <= 30.0 && effort <= 50000.0 {
            return "Fair complexity - consider simplifying some functions";
        }
        "High complexity - code should be refactored for better maintainability"
    }

    /// Emoji assessment for volume (`GetVolumeAssessment`).
    #[must_use]
    pub fn volume_assessment(&self, volume: f64) -> String {
        if volume <= VOLUME_THRESHOLD_HIGH {
            "🟢 Low".to_string()
        } else if volume <= MAGIC_1000 {
            "🟡 Medium".to_string()
        } else {
            "🔴 High".to_string()
        }
    }

    /// Emoji assessment for difficulty (`GetDifficultyAssessment`).
    #[must_use]
    pub fn difficulty_assessment(&self, difficulty: f64) -> String {
        if difficulty <= DIFFICULTY_THRESHOLD_HIGH {
            "🟢 Simple".to_string()
        } else if difficulty <= MAGIC_15 {
            "🟡 Moderate".to_string()
        } else {
            "🔴 Complex".to_string()
        }
    }

    /// Emoji assessment for effort (`GetEffortAssessment`).
    #[must_use]
    pub fn effort_assessment(&self, effort: f64) -> String {
        if effort <= EFFORT_THRESHOLD_HIGH {
            "🟢 Low".to_string()
        } else if effort <= MAGIC_10000 {
            "🟡 Medium".to_string()
        } else {
            "🔴 High".to_string()
        }
    }

    /// Emoji assessment for delivered bugs (`GetBugAssessment`).
    #[must_use]
    pub fn bug_assessment(&self, bugs: f64) -> String {
        if bugs <= BUGS_THRESHOLD_MIN {
            "🟢 Low Risk".to_string()
        } else if bugs <= BUGS_THRESHOLD_MEDIUM {
            "🟡 Medium Risk".to_string()
        } else {
            "🔴 High Risk".to_string()
        }
    }

    /// Serializes the computed metrics as indented JSON (`FormatReportJSON`):
    /// `json.MarshalIndent(metrics, "", "  ")` — two-space indent, HTML escaping
    /// ON, no trailing newline (MarshalIndent does not append one).
    pub fn write_json<W: Write>(&self, metrics: &ComputedMetrics, w: &mut W) -> io::Result<()> {
        let value = metrics.to_go_value();
        let bytes = gojson::Encoder::new()
            .with_indent(Some("  "))
            .with_escape_html(true)
            .with_trailing_newline(false)
            .encode(&value)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("formatreportjson: {e}")))?;
        w.write_all(&bytes)
    }

    /// Serializes the computed metrics as a CFB1 binary envelope
    /// (`FormatReportBinary`): `b"CFB1"` + u32-LE length + compact Go JSON.
    pub fn write_binary<W: Write>(&self, metrics: &ComputedMetrics, w: &mut W) -> io::Result<()> {
        let value = metrics.to_go_value();
        gojson::encode_binary_envelope(&value, w)
    }
}

// NOTE: YAML output (`FormatReportYAML` -> yaml.Marshal) is intentionally NOT
// implemented here. It must route through the shared `cf-goyaml` emitter
// (DESIGN §2.4), which is a bare scaffold at port time. Wiring it in is a
// blocked item tracked in the crate todos; emitting YAML via any other library
// would violate the byte-identity rule.

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn assessments() {
        let f = ReportFormatter::new();
        assert_eq!(f.volume_assessment(100.0), "🟢 Low");
        assert_eq!(f.volume_assessment(6000.0), "🟡 Medium");
        assert_eq!(f.volume_assessment(20000.0), "🔴 High");

        assert_eq!(f.difficulty_assessment(3.0), "🟢 Simple");
        assert_eq!(f.difficulty_assessment(10.0), "🟡 Moderate");
        assert_eq!(f.difficulty_assessment(40.0), "🔴 Complex");

        assert_eq!(f.effort_assessment(500.0), "🟢 Low");
        assert_eq!(f.effort_assessment(5000.0), "🟡 Medium");
        assert_eq!(f.effort_assessment(60000.0), "🔴 High");

        assert_eq!(f.bug_assessment(0.05), "🟢 Low Risk");
        assert_eq!(f.bug_assessment(0.3), "🟡 Medium Risk");
        assert_eq!(f.bug_assessment(1.2), "🔴 High Risk");
    }

    #[test]
    fn message_tiers() {
        let f = ReportFormatter::new();
        assert_eq!(
            f.halstead_message(50.0, 3.0, 500.0),
            "Excellent complexity - code is simple and maintainable"
        );
        assert_eq!(
            f.halstead_message(500.0, 10.0, 5000.0),
            "Good complexity - code is reasonably complex"
        );
        assert_eq!(
            f.halstead_message(3000.0, 25.0, 40000.0),
            "Fair complexity - consider simplifying some functions"
        );
        assert_eq!(
            f.halstead_message(6000.0, 35.0, 60000.0),
            "High complexity - code should be refactored for better maintainability"
        );
    }
}
